service/history/mutableStateBuilder.go

```diff
@@ -699,82 +731,143 @@ func (e *mutableStateBuilder) AddWorkflowExecutionStartedEvent(domainID string,
 	}
 
-	newDecisionEvent := e.hBuilder.AddDecisionTaskScheduledEvent(taskList, startToCloseTimeoutSeconds)
 	di := &decisionInfo{
-		ScheduleID:      *newDecisionEvent.EventId,
+		ScheduleID:      scheduleID,
 		StartedID:       emptyEventID,
 		RequestID:       emptyUUID,
 		DecisionTimeout: startToCloseTimeoutSeconds,
+		Tasklist:        taskList,
+		Attempt:         e.executionInfo.DecisionAttempt,
 	}
 	e.UpdateDecision(di)
 
-	return newDecisionEvent, di
+	return di
 }
 
 func (e *mutableStateBuilder) AddDecisionTaskStartedEvent(scheduleEventID int64, requestID string,
-	request *workflow.PollForDecisionTaskRequest) *workflow.HistoryEvent {
+	request *workflow.PollForDecisionTaskRequest) (*workflow.HistoryEvent, *decisionInfo) {
 	hasPendingDecision := e.HasPendingDecisionTask()
-	pendingDecisionTask, ok := e.GetPendingDecision(scheduleEventID)
-	if !hasPendingDecision || !ok || pendingDecisionTask.StartedID != emptyEventID {
+	di, ok := e.GetPendingDecision(scheduleEventID)
+	if !hasPendingDecision || !ok || di.StartedID != emptyEventID {
 		logging.LogInvalidHistoryActionEvent(e.logger, logging.TagValueActionDecisionTaskStarted, e.GetNextEventID(), fmt.Sprintf(
 			"{HasPending: %v, ScheduleID: %v, Exist: %v, Value: %v}", hasPendingDecision, scheduleEventID, ok, e))
-		return nil
+		return nil, nil
 	}
 
-	event := e.hBuilder.AddDecisionTaskStartedEvent(scheduleEventID, requestID, request)
+	var event *workflow.HistoryEvent
+	scheduleID := di.ScheduleID
+	startedID := scheduleID + 1
+	timestamp := time.Now().UnixNano()
+	// First check to see if new events came since transient decision was scheduled
+	if di.Attempt > 0 && di.ScheduleID != e.GetNextEventID() {
+		// Also create a new DecisionTaskScheduledEvent since new events came in when it was scheduled
+		scheduleEvent := e.hBuilder.AddDecisionTaskScheduledEvent(di.Tasklist, di.DecisionTimeout, 0)
+		scheduleID = scheduleEvent.GetEventId()
+		di.Attempt = 0
+	}
+
+	// Avoid creating new history events when decisions are continuously failing
+	if di.Attempt == 0 {
+		// Now create DecisionTaskStartedEvent
+		event = e.hBuilder.AddDecisionTaskStartedEvent(scheduleID, requestID, request.GetIdentity())
+		startedID = event.GetEventId()
+		timestamp = int64(0)
+	}
 
-	// Update mutable decision state
-	e.executionInfo.DecisionStartedID = *event.EventId
-	e.executionInfo.DecisionRequestID = requestID
 	e.executionInfo.State = persistence.WorkflowStateRunning
+	// Update mutable decision state
+	di = &decisionInfo{
+		ScheduleID:      scheduleID,
+		StartedID:       startedID,
+		RequestID:       requestID,
+		DecisionTimeout: di.DecisionTimeout,
+		Attempt:         di.Attempt,
+		Timestamp:       timestamp,
+	}
+	e.UpdateDecision(di)
 
-	return event
+	return event, di
+}
+
+func (e *mutableStateBuilder) createTransientDecisionEvents(di *decisionInfo, identity string) (*workflow.HistoryEvent,
+	*workflow.HistoryEvent) {
+	tasklist := e.executionInfo.TaskList
```

Comment:
do you need to consider sticky task list here? you might not need to since you already clear sticky when failure happen.

Reply:
Currently I have simplified the state machine to clear sticky tasklist on each failure.  Otherwise you are right we need additional in mutable state to remember that too.
Let's keep it simple until there is a need to remember tasklist also for failed decisions.